src/types/widget.ts

```diff
@@ -6,18 +6,55 @@ import {
   SearchParameters,
 } from './instantsearch';
 
-interface WidgetArgs {
-  helper: Helper;
+interface InitOptions {
+  instantSearchInstance: InstantSearch;
   state?: HelperState;
-  results?: SearchResults;
+  helper?: Helper;
+  templatesConfig?: object;
+  createURL?(state: HelperState): string;
+}
+
+interface RenderOptions {
   instantSearchInstance: InstantSearch;
+  templatesConfig?: object;
+  results?: SearchResults;
+  state?: HelperState;
+  helper?: Helper;
+  searchMetadata?: {
+    isSearchStalled: boolean;
+  };
+  createURL?(state: HelperState): string;
+}
+
+interface DisposeOptions {
+  helper?: Helper;
+  state: HelperState;
 }
 
+type UiState = {
+  [stateKey: string]: any;
+};
+
 export interface Widget {
-  init(options: WidgetArgs): void;
-  render(options: WidgetArgs): void;
-  dispose(options: any): SearchParameters;
-  getConfiguration?(previousConfiguration: SearchParameters): SearchParameters;
+  init?(options: InitOptions): void;
+  render?(options: RenderOptions): void;
+  dispose?(options: DisposeOptions): HelperState;
```

Comment:
The function can return `undefined`.

https://github.com/algolia/instantsearch.js/blob/d553502f13e7eb9b6d423f85983c19e5732cc7a5/src/lib/InstantSearch.js#L222-L229